File: test/python/bench/test_ghz.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the ghz benchmark."""

from __future__ import annotations

import pytest

from mqt.core import bench
from mqt.core.bench import ghz

from .utils import assert_generates


def _make_benchmark() -> ghz.GHZ:
    return ghz.GHZ(
        ghz.Options(
            qubits=3,
            topology=ghz.Topology.STAR,
            basis=ghz.Basis.X,
        )
    )


def test_ghz_options_are_keyword_only_and_read_only() -> None:
    """Keep GHZ options keyword-only and immutable."""
    with pytest.raises(TypeError):
        ghz.Options(3)  # ty: ignore[missing-argument, too-many-positional-arguments]

    options = ghz.Options(
        qubits=3,
        topology=ghz.Topology.STAR,
        basis=ghz.Basis.X,
    )
    with pytest.raises(AttributeError):
        options.qubits = 4  # ty: ignore[invalid-assignment]


def test_ghz_reference() -> None:
    """Expose the output and reference distribution."""
    benchmark = _make_benchmark()
    assert isinstance(benchmark.output, bench.Output)
    assert benchmark.output.name == "result"
    assert benchmark.output.width == 3
    assert benchmark.probability("011") == pytest.approx(0.25)
    assert benchmark.probability("111") == 0


def test_ghz_evaluation() -> None:
    """Evaluate counts against the GHZ reference distribution."""
    benchmark = _make_benchmark()
    evaluation = benchmark.evaluate({"000": 50, "011": 50})
    assert isinstance(evaluation, bench.Evaluation)
    assert evaluation.total_variation_distance == pytest.approx(0.5)
    assert evaluation.squared_hellinger_fidelity == pytest.approx(0.5)
    assert evaluation.success_probability is None


def test_ghz_json_roundtrip() -> None:
    """Preserve the benchmark through both JSON representations."""
    benchmark = _make_benchmark()
    instance_copy = ghz.GHZ.from_instance_specification_json(benchmark.instance_specification_json)
    manifest_copy = ghz.GHZ.from_manifest_json(benchmark.manifest_json)
    assert instance_copy.instance_specification_json == benchmark.instance_specification_json
    assert manifest_copy.manifest_json == benchmark.manifest_json
    assert instance_copy.case_id == manifest_copy.case_id == benchmark.case_id


def test_ghz_generation() -> None:
    """Generate a GHZ program through the Python binding."""
    assert_generates(_make_benchmark().generate())

File: test/bench/test_ghz.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/Evaluation.hpp"
#include "bench/GHZ.hpp"

#include "support/Diagnostics.hpp"
#include "support/TestSupport.hpp"

#include "gtest/gtest.h"

#include <cstddef>
#include <limits>
#include <string>

namespace {

using mqt::bench::Counts;
using mqt::bench::GHZ;
using mqt::bench::GHZBasis;
using mqt::bench::GHZOptions;
using mqt::bench::GHZTopology;
using mqt::bench::Output;

TEST(GHZ, UsesDocumentedDefaults) {
  const auto ghz = ::mqt::test::value(GHZ::create({.qubits = 3}));
  EXPECT_EQ(ghz.options().topology, GHZTopology::Linear);
  EXPECT_EQ(ghz.options().basis, GHZBasis::Z);
  EXPECT_EQ(ghz.output(), (Output{"result", 3}));
}

TEST(GHZ, RejectsUnsupportedQubitCounts) {
  EXPECT_EQ(::mqt::test::errorKind([&] { return GHZ::create({.qubits = 0}); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return GHZ::create({.qubits = GHZOptions::MAX_QUBITS + 1});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_NO_THROW(static_cast<void>(::mqt::test::value(
      GHZ::create({.qubits = GHZOptions::MAX_X_BASIS_QUBITS + 1}))));
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return GHZ::create({.qubits = GHZOptions::MAX_X_BASIS_QUBITS + 1,
                                  .basis = GHZBasis::X});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(GHZ, RejectsUnknownEnumValues) {
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  constexpr auto invalidTopology = static_cast<GHZTopology>(2);
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  constexpr auto invalidBasis = static_cast<GHZBasis>(2);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return GHZ::create({.qubits = 2, .topology = invalidTopology});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return GHZ::create({.qubits = 2, .basis = invalidBasis});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(GHZ, GivesTheZBasisDistribution) {
  const auto ghz = ::mqt::test::value(
      GHZ::create({.qubits = 3, .topology = GHZTopology::Star}));
  EXPECT_DOUBLE_EQ(::mqt::test::value(ghz.probability("000")), 0.5);
  EXPECT_DOUBLE_EQ(::mqt::test::value(ghz.probability("111")), 0.5);
  EXPECT_DOUBLE_EQ(::mqt::test::value(ghz.probability("010")), 0.);
}

TEST(GHZ, GivesTheXBasisDistribution) {
  const auto ghz =
      ::mqt::test::value(GHZ::create({.qubits = 3, .basis = GHZBasis::X}));
  EXPECT_DOUBLE_EQ(::mqt::test::value(ghz.probability("000")), 0.25);
  EXPECT_DOUBLE_EQ(::mqt::test::value(ghz.probability("011")), 0.25);
  EXPECT_DOUBLE_EQ(::mqt::test::value(ghz.probability("111")), 0.);

  const auto largest = ::mqt::test::value(GHZ::create(
      {.qubits = GHZOptions::MAX_X_BASIS_QUBITS, .basis = GHZBasis::X}));
  EXPECT_GT(::mqt::test::value(largest.probability(
                std::string(GHZOptions::MAX_X_BASIS_QUBITS, '0'))),
            0.);
}

TEST(GHZ, EvaluatesCountsAgainstTheWholeIdealDistribution) {
  const auto ghz = ::mqt::test::value(GHZ::create({.qubits = 2}));
  const auto exact = ::mqt::test::value(ghz.evaluate({{"00", 50}, {"11", 50}}));
  EXPECT_DOUBLE_EQ(exact.totalVariationDistance, 0.);
  EXPECT_DOUBLE_EQ(exact.squaredHellingerFidelity, 1.);
  EXPECT_FALSE(exact.successProbability.has_value());

  const auto incomplete = ::mqt::test::value(ghz.evaluate({{"00", 100}}));
  EXPECT_DOUBLE_EQ(incomplete.totalVariationDistance, 0.5);
  EXPECT_DOUBLE_EQ(incomplete.squaredHellingerFidelity, 0.5);
}

TEST(GHZ, ValidatesOutcomesAndShotCounts) {
  const auto ghz = ::mqt::test::value(GHZ::create({.qubits = 2}));
  EXPECT_EQ(::mqt::test::errorKind([&] { return ghz.probability("0"); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] { return ghz.probability("0x"); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] { return ghz.evaluate({}); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] { return ghz.evaluate({{"00", 0}}); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return ghz.evaluate(
                  {{"00", std::numeric_limits<size_t>::max()}, {"11", 1}});
            }),
            ::mqt::ErrorCategory::Overflow);
}

} // namespace
